**#0 Importy**

In [144]:
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

**#1 Wczytanie danych**

In [145]:
data = load_iris()
print(data.keys())

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])


In [146]:
print(data['data'][:5])  # Pierwsze 5 próbek
print(data['feature_names'])  # Nazwy cech
print(data['target'][:5])  # Pierwsze 5 etykiet klas
print(data['target_names'])  # Nazwy gatunków

[[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]]
['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']
[0 0 0 0 0]
['setosa' 'versicolor' 'virginica']


In [147]:
X = data['data'] 
y = data['target']

In [148]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

In [149]:
print(X_train.shape, X_test.shape)

(120, 4) (30, 4)


In [150]:
# Standaryzacja
scaler = StandardScaler()
X_train_standardized = scaler.fit_transform(X_train)
X_test_standardized = scaler.transform(X_test)

**#2 Deklaracja modelu**

In [151]:
model = Sequential([
    Dense(64, activation='relu', input_dim=4),  # 4 cechy wejściowe
    Dense(32, activation='relu'),
    Dense(3, activation='softmax')   # 3 klasy → softmax
])

C:\Users\Mateusz\anaconda3\Lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


**#3 Kompilacja modelu**

In [152]:
adam_opt = tf.keras.optimizers.Adam(learning_rate=0.001)

In [153]:
model.compile(
    optimizer=adam_opt,
    loss="sparse_categorical_crossentropy",  # Bo etykiety są liczbami (0,1,2)
    metrics=["accuracy"]
)

**#4 Podgląd modelu**

In [154]:
model.summary()

Model: "sequential_12"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ dense_36 (Dense)                     │ (None, 64)                  │             320 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_37 (Dense)                     │ (None, 32)                  │           2,080 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_38 (Dense)                     │ (None, 3)                   │              99 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 2,499 (9.76 KB)

 Trainable params: 2,499 (9.76 KB)

 Non-trainable params: 0 (0.00 B)

**#5 Trenowanie modelu**

In [155]:
history = model.fit(
    X_train_standardized, 
    y_train, 
    epochs=50, 
    validation_data=(X_test_standardized, y_test),
)

Epoch 1/50
4/4 ━━━━━━━━━━━━━━━━━━━━ 2s 113ms/step - accuracy: 0.5477 - loss: 1.0173 - val_accuracy: 0.7000 - val_loss: 0.9471
Epoch 2/50
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 31ms/step - accuracy: 0.6994 - loss: 0.9352 - val_accuracy: 0.7667 - val_loss: 0.8836
Epoch 3/50
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 29ms/step - accuracy: 0.7479 - loss: 0.8932 - val_accuracy: 0.7667 - val_loss: 0.8220
Epoch 4/50
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 28ms/step - accuracy: 0.8029 - loss: 0.8331 - val_accuracy: 0.7667 - val_loss: 0.7628
Epoch 5/50
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 27ms/step - accuracy: 0.8185 - loss: 0.7422 - val_accuracy: 0.8000 - val_loss: 0.7083
Epoch 6/50
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 29ms/step - accuracy: 0.8415 - loss: 0.7031 - val_accuracy: 0.8000 - val_loss: 0.6590
Epoch 7/50
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 34ms/step - accuracy: 0.8017 - loss: 0.6542 - val_accuracy: 0.8000 - val_loss: 0.6149
Epoch 8/50
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 32ms/step - accuracy: 0.8223 - loss: 0.5789 - val_accuracy: 0.8000 - val_loss: 0.5754

**#6 Ocena modelu**

In [156]:
test_loss, test_acc = model.evaluate(X_test_standardized, y_test)
print(f"Dokładność na zbiorze testowym: {test_acc:.2f}")

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 74ms/step - accuracy: 0.9333 - loss: 0.1279
Dokładność na zbiorze testowym: 0.93


**#7 Podsumowanie**

Podsumowując, dokładność 93 % jest bardzo dobrą wartością dla klasyfikacji Iris, co oznacza, że model radzi sobie świetnie w rozróżnianiu gatunków kwiatów na podstawie ich cech.